common/src/main/java/io/pravega/common/util/ContinuationTokenAsyncIterator.java

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.common.util;
+
+import com.google.common.annotations.VisibleForTesting;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+
+@ThreadSafe
+@Slf4j
+/**
+ * This is a continuation token based async iterator implementation. This class takes a function that when completed will 
+ * have next batch of results with continuation token. 
+ * This class determines when to call the next iteration of function (if all existing results have been exhausted) and 
+ * ensures there is only one outstanding call. 
+ */
+public class ContinuationTokenAsyncIterator<T, U> implements AsyncIterator<U> {
+    private final Object lock = new Object();
+
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final Queue<U> queue;
+    @GuardedBy("lock")
+    @VisibleForTesting
+    @Getter(AccessLevel.PACKAGE)
+    private final AtomicReference<T> token;
+    private final Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function;
+    @GuardedBy("lock")
+    private CompletableFuture<Void> outstanding;
+    private final AtomicBoolean canContinue;
+    
+    public ContinuationTokenAsyncIterator(Function<T, CompletableFuture<Map.Entry<T, Collection<U>>>> function, T tokenIdentity) {
+        this.function = function;
+        this.token = new AtomicReference<>(tokenIdentity);
+        this.queue = new LinkedBlockingQueue<>();
+        this.outstanding = CompletableFuture.completedFuture(null);
+        this.canContinue = new AtomicBoolean(true);
+    }
+
+    @Override
+    public CompletableFuture<U> getNext() {
+        final T continuationToken;
+        synchronized (lock) {
+            // if the result is available, return it without making function call
+            if (!queue.isEmpty()) {
+                return CompletableFuture.completedFuture(queue.poll());
+            } else {
+                continuationToken = token.get();
+                // make the function call if previous outstanding call completed.
+                if (outstanding.isDone()) {
+                    // We are making this call under the lock!
```

Comment:
Ok, now I understand what you meant in your first response. It does look odd to hold a lock while making this call given that we are returning a future. In your previous implementation, you were not duplicating elements, but possibly duplicating calls for a given continuation token. My concern was with the high-level semantics of the iterator and not so much the efficiency of the implementation.

My preference is that we do not hold the lock for this call, but at the same time, I think we should return a future without waiting on this call. Would it be possible?

If this is getting unnecessarily complicated, we can perhaps create an issue to optimize this later.

Reply:
One of the hacky alternatives i could think of was something along these lines:
```
                    outstanding = CompletableFuture.runAsync(() -> {}).thenCompose(x -> function.apply(continuationToken)
                            .thenAccept(resultPair -> {
```

This hack will return a future immediately while running the chained callbacks on an external thread.
If we had our executor, we could have scheduled this on our executor. and in its absence it will go to forkjoinpool. 
But it takes ticks all the boxes.. does not make multiple remote calls for same continuation token.. neither does it hold the lock while making a remote call.

Reply:
I think there is a simple way we can achieve what we desire.. updating the code. Please take a look